### ISSUE TITLE
Using Live Design inside an IDE

### ISSUE THREAD

**Issue by coolbluewater** [makepad/makepad#207]
Hi, the documentation says that:

> The Live system enables live coding by providing the machinery for an IDE to:
> - connect to a running application
> - detect whether the code being edited is part of a Live DSL block

How to do this from an IDE other than makepad, VSCode in my case?

**Comment by makepaddev** [comment]
Currently i use filewatchers inside any makepad application on its own sourcecode. If you set vscode to autosave (or save manually) it will hotload the DSL.
HOWEVER this is extremely new and barely tested and will have lots of edgecase failures still of something not updating properly, but you can try it.

**Comment by makepaddev** [comment]
Such as right now the shaders don't hotload which they should, need to fix that up.

**Comment by coolbluewater** [comment]
Got it. Running into a weird bug while trying this, will open an issue.

**Comment by makepaddev** [comment]
Yea its really really untested so don't put much faith in it just yet. I just barely got it to work.

**Comment by coolbluewater** [comment]
Speaking of testing, I don't see any documentation or code related to testability. Is this not part of makepad's architecture?

**Comment by makepaddev** [comment]
The plan is to use a form of headless automated testing that plugs straight into the system. Just like i run an application 'cross process' at the moment you can make a platform layer that exists only for testing applications and has a deterministic input/output strategy. We havent built that yet tho since every platform i add is weight. Also we haven't needed much yet as our primary cycle is dogfooding, rust catches a ton of issues anyway, and a lot of things are very hard to test because they are platform specific.

**Comment by makepaddev** [comment]
This is also where the accessibility apis (nav trees) will fit in, i want to write tests against those apis and hten look at the render command output to write decently runnable tests. All of this is a todo item tho.

**Comment by makepaddev** [comment]
As for the entire DSL system i wrote some tests for that in the beginning but just keeping the tests working proved to be such a drag on the evolutionspeed of it i gave up. Once it stabilises we can 'lock' it with a proper set of tests. For now the applications we write are the tests.

**Comment by makepaddev** [comment]
Other naming suggestions for Turtle Walk and Layout are welcome if you have better names. However it needs to be significantly better for me to change it everywhere because it'l trip up our own team badly.

**Comment by makepaddev** [comment]
Including myself. Every time i rename something my brain needs to reconfigure so i slow myself down.

**Comment by coolbluewater** [comment]
> Once it stabilises we can 'lock' it with a proper set of tests. For now the applications we write are the tests.
This is pretty weak and imho reverse thinking. Contrary to your belief that this adds weight, tests driven development guides the API design and usability and stability of the product and shaves years off the development cycle. 
Rust isn't a magic bullet for the kind of issues I'm already seeing. I mean, how is Rust going to know that typing a character should not close the document?
I've spent *years* writing tests for editors and have not regretted a minute of it. This is the reason imho why the development of makepad is proceeding as it is - working blind with no real insight into what functionality works what not.

Again, this is not an attack. It is simply a prediction that needs this uncompromising language to communicate - without automated tests at every level the project is doomed. There's no modern successful project that I can think of that has gotten away without this.

And on the flip side, once testing becomes an ingrained methodology suddenly things feel like they are under control again. Suddenly the panic disappears and engineering takes on a predictable cadence. 

I don't know if this response will have any effect, sadly I suspect it won't.

**Comment by makepaddev** [comment]
I see your point, however i've also experienced evolving this stack with intense testing and it just doesn't work. Im not sure you realise how much our systems have evolved over the last few years. A year ago we didnt even HAVE a UI design DSL. I can spend all that time iterating my tests, or iterate the system.

**Comment by makepaddev** [comment]
We do have tests for supporting libraries we develop btw.

**Comment by makepaddev** [comment]
Makepad is not a straightforward design and implement project, its been 90% exploratory of how to even do this at all. Similarly our layout engine started with a premise of: Make an immediate mode layout engine that does not need a layout-pass. And evolve from there. I do think that thing needs to get some serious tests soon and possibly a reimplementation by my co-founder that puts it on more sound ground in terms of mathematical defs.

**Comment by makepaddev** [comment]
Because yes at some point without tests you will feel like yr blind, and generally be panicky about changing anything and then you grind to a halt in that way.

**Comment by coolbluewater** [comment]
> I see your point, however i've also experienced evolving this stack with intense testing and it just doesn't work.

That's because your "it" is obviously not the "it" I'm referring to. Testing is another area that needs serious study. At some point I understood this and took time out to delve into the field and to my utter amazement discovered a world of unsuspected riches.

**Comment by makepaddev** [comment]
It's a difference of opinion how to iterate/evolve quickly when yr exploring. After you're done exploring you need to start testing, thats how i view it. And that moment is arriving soon once i have a visual designtool that actually works (which is not at all a given yet)

**Comment by coolbluewater** [comment]
Yes, as already stated, I didn't think this would have any effect.

**Comment by makepaddev** [comment]
Your input is processed tho.

**Comment by coolbluewater** [comment]
Not a worthy comment, sorry.

**Comment by coolbluewater** [comment]
An assessment of makepad has been forming for some time. 

At this point makepad feels like a videogame that's based around a single game mechanic or interesting hack. There have been videogames in the past that succeeded because their developers found an interesting way to overclock the hardware, such as by displaying more sprites or colors than normally thought possible or animate using the color table.

In makepad that interesting hack is displaying a vector tree via the GPU in a cross-platform way. That's it. But the price to entry is you're stuck in the makepad walled garden. *Everything* needs to be built using this hack, even the parts that ought not to require it. There's been no thought about how to create real apps. The menubar on mac is empty. Command-Q does not even work to close the app - after years of makepad in its current incarnation. 

I mean, it's so easy to just *imagine* stuff. In the shower. While taking a walk. Then the scenarios and questions come naturally. "What does we need to be a GUI toolkit"? "What does it mean to be an IDE"? What about extensions, keybindings, LSP support, integrated terminals, integrated Jupyter notebooks, AI code tools? And these are just the known ways people interact with code. Do you guys really expect people to ditch all that for animated code folding? 

None of this should overwhelm. But without the actual practice of being able to stabilize the mind and use it regularly and meditatively to create extraordinary worlds within, nothing extraordinary will emerge in the one without.

The handwaving response at conference to questions about particular features, that they are easily possible, that you've just not gotten around to them feels irresponsible. Like other editors and GUI stacks (Xi and druid come to mind) you've fallen into the illusion that things are orders of magnitude easier than they actually are, instead of just admitting when things have just not been thought through and leaving it at that. And also the constant refrain of building a "full UI stack" without even admitting to yourself that this is years and years into the future at the current trajectory and methodology.

The "live design" is founded on a false premise - that only style needs to be live edited and that it is not possible to live edit code changes. This is false, and based on not keeping up with competing technologies that have figured out how to do this via an intimate understanding of compilers and linkers.

I am fully aware of the high quality of this assessment, and my own technical prowess at this point. So feel free to ignore it altogether, but please avoid patronizing responses. I don't need to be consoled - this is your project and your life-times. I merely check in every so often and take a peek, and never felt the need to voice an opinion until now.

**Comment by makepaddev** [comment]
Hey i don't disagree with your assesment. However having also spent more than a decade in trying to build UI in a browser using the 'standards' that people made, i have given up and embarked on creating a new path. This trick you mention is indeed a trick. But one that enables us to create working UI that isn't heavily bottlenecked by some huge C++ render engine in the back. And yes that comes with compromises, but it is a compromise that has enabled me personally to build what i wanted to build, where the browser stack has mostly failed every single attempt to push things forward.
Now we intend to flesh this out fully for others to use, and others are using it. But it is indeed a long road. But we're walking it. We do have mac menu support, its in there. I just need to enable it again. We are building out in the open, and that comes with these consequences of 'yea i havent integrated it back in yet'. Thanks for your feedback tho and taking the time to have a look. I'm sure we'll be closer to checking the boxes you mentioned in the future.

**Comment by makepaddev** [comment]
Also live design is not a false premise. It is a compromise. A compromise that most if not all of other Rust UI stacks don't even make at all. And if you integrate a recompile cycle 'smoothly' in your IDE as we are doing, you can approach live-coding from that side as well. But i'll take an optimized compromise over just giving up entirely. And that goes for the rest of the UI stack as well. It's not perfect, its a compromise. But it is one that gets me, personally, a hell of a lot closer to where i want to be than nothing, and i hope others as well over time.

**Comment by makepaddev** [comment]
If you are referring to 'Dart and flutter' Yes they have figured out a more optimised way to live code. It is also a heavyweight VM and an enormous C++ stack. I have yet to run into people claiming they can build anything in Dart such as a synth or any kind of compute heavy code. You have to drop immediately to C++ or Rust to support that. So yay you get livecoding, but yr still stuck in a dynamic language VM so. Dart is its own compromise. And i'm happy i got Rust this time around and not JS or something like that.

**Comment by makepaddev** [comment]
And the list of todo items for the IDE Like LSP and keybindings and so forth, Yep the list is long. But we're going to get through it regardless. I've tried to build an IDE company on HTML and browser tech before and it was a complete nightmare. With makepad we're already at the point now where i can dogfood (or atleast getting back to that point soon). So again, for me personally that is a huge milestone. And over time i hope that others can use it too.

**Comment by coolbluewater** [comment]
> If you are referring to 'Dart and flutter' 
Nope, not dart or flutter. I purposely avoided mentioning the technologies I'm thinking about to make it an adventure of discovery for you. I hope you are as shocked and amazed as I've been when & if you stumble upon them.

Ultimately, in this game everything counts. To build the kind of thing you're building it helps to be a bit of a compiler guru (or to learn from them).  I'm not suggesting you drop your current design, but am suggesting that your core premise is likely to be blindsided in some near future.

**Comment by makepaddev** [comment]
Very well possible, and indeed this is part of the stack that could in an ideal world be a LOT better. But i dont have the resources to make my own compiler or language. Rust is far from perfect, but its a LOT better than C++ or JS or Zig so i'll take it.

**Comment by coolbluewater** [comment]
I've nowhere referred to making a new language :)

**Comment by makepaddev** [comment]
If you want me to comment in any way that pertains to what you know and i dont you need to atleast point me to something i can read.

**Comment by coolbluewater** [comment]
No comment needed on that topic :)

> We do have mac menu support, its in there. I just need to enable it again.

These are the kinds of things that leave a great impression on people. 
The README for example, is worth about five orders of magnitude more, pound for pound, than any code file.
Keeping things working and tidy and not crashing or misbehaving while revamping the house internally with a jackhammer is not easy but well worth it. I mean makepad *looks* that way in screenshots, but sadly does not work that way in practice. (Ironfish is much better.)

**Comment by makepaddev** [comment]
Makepad 'studio' is a week old and barely works, but yea i do agree our communication strategy really is not good enough at this point in the attention cycle.

**Comment by makepaddev** [comment]
Building out in public as we do has mostly worked fine, but sometimes people come in with the wrong expectations for where we are.

**Comment by coolbluewater** [comment]
I mean sorry to sound offensive but I've taken plenty of alpha dev tools for a spin and makepad studio has always been at the bottom of the barrel. Really sorry but I know you like to hear the truth.

**Comment by makepaddev** [comment]
Makepad studio has also not been a devtool at all yet. I cant even use it myself. So i'm not surprised. However we should probably be very explicit about communicating that.

**Comment by coolbluewater** [comment]
I mean whatever name this IDE has been going under for the last three years.

**Comment by makepaddev** [comment]
I made a prototype very early on that im using myself teeth grinding, and then we rewrote the stack for about 3 years untill we had a DSL UI stack / Rust API's i can actually make a settings UI with, and has atleast the idea working of a visual designtool.  We are about 2 weeks in where i have put together a rudimentary version of something you can type in and it runs a childprocess in a shared texture on the side. So anyway. communication.

**Comment by makepaddev** [comment]
I am expecting things to accelerate now since all the bits are sorta working well enough now to finally build our IDE.

**Comment by coolbluewater** [comment]
Glad to hear it. Gloomy prediction - this is the fun the period in which things move fast because it's all in your head. Then you start to forget things and if it hasn't been built with TDD things fall apart and it becomes a pain to wake up each day.

**Comment by makepaddev** [comment]
It's definitely an issue we'll tackle soon yea because that moment will come.

**Comment by coolbluewater** [comment]
In the meantime I've decided to take a bit of time to experiment with this cool stack you've built for some custom UI. Is there a way currently to embed an instance within a view of a normal mac app? If not, no worries I'll figure it out.

**Comment by makepaddev** [comment]
I mean its all NSViews and the code is right there. So if you can reparent an NSView you should be able to do it.

**Comment by coolbluewater** [comment]
Is that all - I mean the macros take over the entire app and runloop, no?

**Comment by makepaddev** [comment]
https://github.com/makepad/makepad/tree/master/platform/src/os/apple
You will see cocoa_app and cocoa_window.rs where i create and handle it.

**Comment by makepaddev** [comment]
Yea the macro creates the entrypoint for the eventloop since it is different for wasm/android/native. the first two get a message-entry point and the native one is a non returning function

**Comment by coolbluewater** [comment]
No worries, this part I can figure out. My question had to do with whether I'd need to modify the makepad repo itself, and looks like the answer is yes.

**Comment by makepaddev** [comment]
its in platform/src/app_main.rs

**Comment by makepaddev** [comment]
Yes. we are at that point in the process. You will also find that we arent really recent on cargo or at all.

**Comment by coolbluewater** [comment]
I did notice 2018. BTW check out rust workspace inheritance in 2021, makes Cargo.toml a lot nicer.

**Comment by makepaddev** [comment]
Ah interesting

**Comment by coolbluewater** [comment]
Maybe I can do this as part of getting started. Interested in a PR that modernizes the Cargo.toml's?

**Comment by makepaddev** [comment]
Sounds great yes thanks.

**Comment by univerz** [comment]
> I purposely avoided mentioning the technologies I'm thinking about to make it an adventure of discovery for you. I hope you are as shocked and amazed as I've been when & if you stumble upon them.

I'm sure we all can be shocked and amazed even if we skip the research stage and study what you find usable.

**Comment by yulnr** [comment]
Closing as this turned into a general conversation. Live design hot reload works through file watchers so it's IDE independent.
